dask/dataframe/groupby.py

```diff
@@ -132,6 +136,304 @@ def _nunique_series_aggregate(df):
     return df.groupby(df.columns[1])[df.columns[0]].nunique()
 
 
+###############################################################
+# Aggregate support
+#
+# Aggregate is implemented as:
+#
+# 1. group-by-aggregate all partitions into intermediate values
+# 2. collect all partitions into a single partition
+# 3. group-by-aggregate the result into intermediate values
+# 4. transform all intermediate values into the result
+#
+# In Step 1 and 3 the dataframe is grouped on the same columns.
+#
+###############################################################
+
+def _make_agg_id_factory():
+    """
+    Return a factory for ids in the aggregate graph.
+
+    Keys are built from the function, the column to apply the function to, and
+    a increasing numeric id. The returns ids are deterministic. However, the
+    order in which ids are generated changes the numeric id. For different
+    calls with the same arguments, the factory function will always return the
+    same id.
+    """
+    integers = it.count()
+    ids = {}
+
+    def factory(func, column):
+        try:
+            return ids[func,column]
+
+        except KeyError:
+            new_id = '{}‐{!s}‐{!s}'.format(next(integers), func, column)
+            return ids.setdefault((func,column), new_id)
+
+    return factory
+
+
+def _aggregate_meta(df, index, levels, chunk_funcs, agg_funcs, finalizers):
+    stage1 = _groupby_apply_funcs(df, funcs=chunk_funcs, by=index)
+    stage2 = _groupby_apply_funcs(stage1, funcs=agg_funcs, level=levels)
+    return _agg_finalize(stage2, finalizers)
+
+
+def _normalize_spec(spec, non_group_columns):
+    """
+    Return a list of ``(result_column, func, input_column)`` tuples.
```

Comment:
Can I ask for a quick `Examples` section here with a couple of worked examples to help future developers quickly understand how this function operates.  

```
Examples
--------

>>> _normalize_spec(simple, inputs)
... simple outputs ...
```